Expanded the article some, and included the diagram. (#1077)
temporalio/documentation: docs/concepts/what-is-an-activity-definition.md

  - explanation
---

An Activity Definition is the set of constraints that apply to an [Activity Execution](/docs/concepts/what-is-an-activity-execution). Activity Definitions are referred to by their [Activity Type](/docs/concepts/what-is-an-activity-type).

> In Temporal documentation, the term 'Activity Definition' will be used to refer to an Activity Function Definition——the method or function that is invoked for an [Activity Task Execution](/docs/concepts/what-is-an-activity-task-execution). Some SDKs refer to Activity Definitions as Activity Functions, Activity Interfaces, or Activity Methods. All of these terms refer to the source of an instance of an execution, with an Activity Interface being a collective group of functions.

Reviewer: ```suggestion
In Temporal documentation, the term 'Activity Definition' is used to refer to an Activity Function Definition——the method or function that is invoked for an [Activity Task Execution](/docs/concepts/what-is-an-activity-task-execution).
Some SDKs refer to Activity Definitions as Activity Functions, Activity Interfaces, or Activity Methods.
All of these terms refer to the source of an instance of an execution, with an Activity Interface being a collective group of functions.
```

> ![Activity Definition](/static/diagrams/activity-definition.svg)

Reviewer: ```suggestion
![Activity Definition](/diagrams/activity-definition.svg)
```

> ## Constraints

Reviewer: Lets use h4s here.

> Temporal does not recover Activity state in the event of failure. Therefore, an Activity Definition has no restrictions on the code it contains.

Reviewer: I would double down on this message to make it clear.

Activity functions are executed as normal functions.
In the event of failure, if the function is executed again, it begins at initial state (with the exception of when Activity Heartbeats are involved).
Therefore - there is no restrictions on the code -

## Parameters

> An Activity Definition can support as many parameters as needed. All values passed through these parameters are recorded in the [Event History](/docs/concepts/what-is-an-event-history).

Reviewer: ```suggestion
An Activity Definition can support as many parameters as needed.
All values passed through these parameters are recorded in the calling Workflow Execution's [Event History](/docs/concepts/what-is-an-event-history).
```

Reviewer 2: "All values passed through these parameters are recorded in the Workflow Execution's [Event History](...)"

> An Activity Definition can support as many parameters as needed. All values passed through these parameters are recorded in the [Event History](/docs/concepts/what-is-an-event-history).

Reviewer: Return values are also captured in the Event History for the calling Workflow Execution.

  - explanation
---

> An Activity Definition is the set of constraints that apply to an [Activity Execution](/docs/concepts/what-is-an-activity-execution). Activity Definitions are referred to by their [Activity Type](/docs/concepts/what-is-an-activity-type.md).

Reviewer: ```suggestion
An Activity Definition is the set of constraints that apply to an [Activity Execution](/docs/concepts/what-is-an-activity-execution). Activity Definitions are referred to by their [Activity Type](/docs/concepts/what-is-an-activity-type).
```

> In Temporal documentation, the term 'Activity Definition' will be used to refer to an Activity Function Definition——the method or function that is invoked for an [Activity Task Execution](/docs/concepts/what-is-an-activity-task-execution.md). Some SDKs refer to Activity Definitions as Activity Functions, Activity Interfaces, or Activity Methods. All of these terms refer to the source of an instance of an execution, with an Activity Interface being a collective group of functions.

Reviewer: ```suggestion
In Temporal documentation, the term 'Activity Definition' will be used to refer to an Activity Function Definition——the method or function that is invoked for an [Activity Task Execution](/docs/concepts/what-is-an-activity-task-execution). Some SDKs refer to Activity Definitions as Activity Functions, Activity Interfaces, or Activity Methods. All of these terms refer to the source of an instance of an execution, with an Activity Interface being a collective group of functions.
```

## Constraints

Temporal does not recover Activity state in the event of failure. Therefore, an Activity Definition has no restrictions on the code it contains.

## Parameters

An Activity Definition can support as many parameters as needed. All values passed through these parameters are recorded in the [Event History](/docs/concepts/what-is-an-event-history.md).

Activity Definitions must contain the following parameters:

- Context: an optional parameter that provides Activity context within multiple APIs.
- Heartbeat: a notification from the Worker to the Temporal Cluster that the Activity Execution is progressing. Cancelations are only allowed if the Activity Definition permits Heartbeating.
- Timeouts: intervals that control the execution and retrying of Activity Task Executions.

Other parameters, such as retry policies and return values, can be seen in the implementation guides below.

## Implementing Activity Definitions

We strongly recommend that you develop an Activity Definition in a language with a corresponding Temporal SDK.

**Implementation guides:**

- [How to develop an Activity Definition in Go](/docs/go/how-to-develop-an-activity-definition-in-go)

> - [How to develop an Activity Interface in Java](/docs/java/activities.md#activity-interface)
> - [How to develop an Activity Interface in PHP](/docs/php/activities.md#activity-interface)
> - [How to develop an Activity Interface in TypeScript](/docs/typescript/activities.md#how-to-write-an-activity-function)

Reviewer: ```suggestion
- [How to develop an Activity Interface in Java](/docs/java/activities/#activity-interface)
- [How to develop an Activity Interface in PHP](/docs/php/activities/#activity-interface)
- [How to develop an Activity Interface in TypeScript](/docs/typescript/activities/#how-to-write-an-activity-function)
```

An Activity Definition is the code that defines the constraints of an [Activity Task Execution](/docs/concepts/what-is-an-activity-task-execution).

Activity Definitions are referred to by their [Activity Type](/docs/concepts/what-is-an-activity-type).

In Temporal documentation, the term 'Activity Definition' is used to refer to an Activity Function Definition——the method or function that is invoked for an [Activity Task Execution](/docs/concepts/what-is-an-activity-task-execution).

> Some SDKs refer to Activity Definitions as Activity Functions, Activity Interfaces, or Activity Methods.

Reviewer 3: ```suggestion
Temporal SDKs refer to Activity Definitions as Activity Functions, Activity Interfaces, or Activity Methods.
```

All of these terms refer to the source of an instance of an execution, with an Activity Interface being a collective group of functions.

![Activity Definition](/diagrams/activity-definition.svg)

#### Constraints

Activity Definitions are executed as normal functions.

> In the event of failure, the function will begin at its initial state when retried (except when Activity Heartbeats are established).

Reviewer 3: ```suggestion
In the event of failure, the function begins at its initial state when retried (except when Activity Heartbeats are established).
```

Per https://developers.google.com/style/tense.

Therefore, an Activity Definition has no restrictions on the code it contains.

#### Parameters

An Activity Definition can support as many parameters as needed.

> All values passed through these parameters are recorded in the Workflow Execution's [Event History](/docs/concepts/what-is-an-event-history). Return values are also captured in the Event History for the calling Workflow Execution.

Reviewer 3: ```suggestion
All values passed through these parameters are recorded in the [Event History](/docs/concepts/what-is-an-event-history) of the Workflow Execution.
Return values are also captured in the Event History for the calling Workflow Execution.
```

First sentence tweaked to avoid possessive and thus provide easier translation later.

Activity Definitions must contain the following parameters:

- Context: an optional parameter that provides Activity context within multiple APIs.

> - Heartbeat: a notification from the Worker to the Temporal Cluster that the Activity Execution is progressing. Cancelations are only allowed if the Activity Definition permits Heartbeating.

Reviewer 3: ```suggestion
- Heartbeat: a notification from the Worker to the Temporal Cluster that the Activity Execution is progressing. Cancelations are allowed only if the Activity Definition permits Heartbeating.
```

- Timeouts: intervals that control the execution and retrying of Activity Task Executions.

> Other parameters, such as retry policies and return values, can be seen in the implementation guides below.

Reviewer 3: ```suggestion
Other parameters, such as [Retry Policies](/docs/concepts/what-is-a-retry-policy) and return values, can be seen in the implementation guides, listed in the next section.
```

Replaced "below" per "below" in https://developers.google.com/style/word-list#letter-b.

#### Implementing Activity Definitions

> We strongly recommend that you develop an Activity Definition in a language with a corresponding Temporal SDK.

Reviewer 3: ```suggestion
We strongly recommend that you develop an Activity Definition in a language that has a corresponding Temporal SDK.
```

An Activity Definition is the code that defines the constraints of an [Activity Task Execution](/docs/concepts/what-is-an-activity-task-execution).

> Activity Definitions are named and referenced in code by their [Activity Type](/docs/concepts/what-is-an-activity-type).
> 
> In Temporal documentation, the term 'Activity Definition' is used to refer to an Activity Function Definition——the method or function that is invoked for an [Activity Task Execution](/docs/concepts/what-is-an-activity-task-execution).
> 
> Temporal SDKs refer to Activity Definitions as Activity Functions, Activity Interfaces, or Activity Methods.
> All of these terms refer to the source of an instance of an execution, with an Activity Interface being a collective group of functions.

Reviewer: ```suggestion
The term 'Activity Definition' is used to refer to the full set of primitives in any given language SDK that provides an access point to an Activity Function Definition——the method or function that is invoked for an [Activity Task Execution](/docs/concepts/what-is-an-activity-task-execution).
Therefore, the terms Activity Function and Activity Method refer to the source of an instance of an execution.

Activity Definitions are named and referenced in code by their [Activity Type](/docs/concepts/what-is-an-activity-type).
```

I would drop the bit about the Activity Interface - for now.

Return values are also captured in the Event History for the calling Workflow Execution.

Activity Definitions must contain the following parameters:

- Context: an optional parameter that provides Activity context within multiple APIs.
- Heartbeat: a notification from the Worker to the Temporal Cluster that the Activity Execution is progressing. Cancelations are allowed only if the Activity Definition permits Heartbeating.
- Timeouts: intervals that control the execution and retrying of Activity Task Executions.

Other parameters, such as [Retry Policies](/docs/concepts/what-is-a-retry-policy) and return values, can be seen in the implementation guides, listed in the next section.

#### Implementing Activity Definitions

We strongly recommend that you develop an Activity Definition in a language that has a corresponding Temporal SDK.

**Implementation guides:**

- [How to develop an Activity Definition in Go](/docs/go/how-to-develop-an-activity-definition-in-go)
- [How to develop an Activity Interface in Java](/docs/java/activities/#activity-interface)
- [How to develop an Activity Interface in PHP](/docs/php/activities/#activity-interface)

> - [How to develop an Activity Interface in TypeScript](/docs/typescript/activities/#how-to-write-an-activity-function)

Reviewer: Can we move the "how-to" links to be after the first intro paragraph/sentence?

We can keep the "corresponding sdk context", but drop the title and the bold text with colon.

Reviewer: actually ignore this comment for now.
